### PR TITLE
[inductor] Fix bug with contiguous checks and comprehensive_padding

### LINKED ISSUE #180848: `torch.compile` produces wrong results when fusing `adaptive_avg_pool2d` with `flatten + sum`
## Describe the bug

`torch.compile(backend='inductor')` silently produces incorrect results when `adaptive_avg_pool2d` is fused with downstream `flatten` + `sum` operations. The bug triggers when the **channel dimension is not a power of 2** (e.g., C=33, 65, 129).

This is a **silent correctness bug** — no error is raised, but the fused kernel computes wrong output values. The `adaptive_avg_pool2d` kernel alone produces correct results; the error is introduced during fusion with the subsequent reduction.

## Minimal repro

```python
import torch

def f(x):
    y = torch.nn.functional.adaptive_avg_pool2d(x, 7)
    return y.flatten(1).sum(dim=-1)

torch.manual_seed(42)
x = torch.randn(2, 33, 8, 8, dtype=torch.float64, device='cuda')

# Eager — correct
print(f"Eager: {f(x).tolist()}")
# [-52.27041819362077, 30.75847231895647]

# Inductor — wrong
torch._dynamo.reset()
compiled = torch.compile(f, backend='inductor')(x)
print(f"Inductor: {compiled.tolist()}")
# [-52.27041819362077, 34.325837206874446]  ← second element is wrong

print(f"Max diff: {(f(x) - compiled).abs().max().item()}")
# 3.567365  ← catastrophic for fp64
```

## Key observations

**1. Pool alone is correct; fusion introduces the error:**

```python
# These are all correct (diff ≈ 0):
torch.compile(lambda x: F.adaptive_avg_pool2d(x, 7))(x)           # pool only
torch.compile(lambda x: F.adaptive_avg_pool2d(x, 7).sum())(x)     # pool + sum
torch.compile(lambda x: F.adaptive_avg_pool2d(x, 7).flatten(1))(x) # pool + flatten

# This is WRONG (diff ≈ 3.6):
torch.compile(lambda x: F.adaptive_avg_pool2d(x, 7).flatten(1).sum(-1))(x)
```

**2. Channel dimension determines whether the bug triggers:**

| C (channels) | Power of 2? | Diff |
|:---|:---|:---|
| 32 | yes | 0 |
| **33** | **no** | **7.37** |
| 48 | no (but 16×3) | 0 |
| 64 | yes | 0 |
| **65** | **no** | **4.71** |
| 128 | yes | 0 |
| **129** | **no** | **6.45** |

**3. Spatial dimensions also matter:**

With C=33, output_size=7:
- H=7 (no resize needed): diff=0 ✅
- H=8 (8→7, non-integer stride): diff=7.37 ❌
- H=14 (14→7, integer stride): diff=0 ✅ on pt211, diff≠0 on nightly
- H=15 (15→7, non-integer stride): diff=8.86 ❌

**4. CUDA only — CPU produces correct results.**

**5. Deterministic across random seeds** — not a race condition.

### Versions

- **PyTorch 2.11.0+cu126**: bug present
- **PyTorch 2.12.0.dev20260410+cu126 (nightly)**: bug present
- CUDA 12.6, Tesla T4
- OS: Ubuntu 20.04, Python 3.11

cc @ptrblck @msaroufim @eqy @jerryzh168 @tinglvv @nWEIdia @mikaylagawarecki @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### LINKED ISSUE #180956: `torch.compile` produces wrong results for `adaptive_avg_pool2d` + `flatten` + `sum` fusion
### 🐛 Describe the bug

`torch.compile(backend='inductor')` silently produces wrong results when `adaptive_avg_pool2d` is fused with `flatten` + `sum` (or `mean`). The bug triggers when **batch > 1** and the flattened reduction dimension (`C × pool_h × pool_w`) exceeds 1024 with channels **not** a multiple of 16.

This is a **silent correctness bug** — no error is raised, the output is simply wrong. The error is large (up to `1e5` in `max-autotune` mode) and persists in `float64`, ruling out precision noise.

## Minimal repro

```python
import torch

x = torch.randn(3, 33, 8, 8, dtype=torch.float64, device='cuda')

def f(x):
    return torch.nn.functional.adaptive_avg_pool2d(x, 7).flatten(1).sum(dim=-1)

# Eager — correct
print(f"eager = {f(x).tolist()}")

# Inductor — wrong
torch._dynamo.reset()
compiled = torch.compile(f, backend='inductor')(x)
print(f"inductor = {compiled.tolist()}")
print(f"max_diff = {(f(x) - compiled).abs().max().item():.2e}")
# max_diff ~ 5e+00 to 8e+01 depending on seed
```

## Expected behavior

Inductor output should match eager output. `adaptive_avg_pool2d(x, 7).flatten(1).sum(dim=-1)` is a standard global-pool-then-reduce pattern used in CNN classifiers.

## Trigger conditions

| Condition | Required? | Notes |
|-----------|-----------|-------|
| `batch > 1` | Yes | `batch=1` always correct; error grows with batch |
| `C % 16 != 0` | Yes | C=16,32,64,128,256,512 all safe |
| `C × pool_h × pool_w > 1024` | Yes | e.g. C=21, pool=7 → 21×49=1029 triggers; C=20 → 980 safe |
| `adaptive_avg_pool2d` fused with `flatten+sum` | Yes | Pool-only or flatten+sum-only are both correct separately |

## Root cause analysis

The bug is **not** in `adaptive_avg_pool2d` itself — pool-only output is correct. It occurs when inductor fuses the pool with `flatten` + `sum` into a single reduction kernel. The tiling/indexing for the fused kernel appears to miscalculate offsets for batch elements beyond the first (batch[0] is always correct, batch[1+] are wrong).

The alignment to multiples of 16 and the 1024-element threshold suggest a Triton reduction tile size boundary issue in the fused kernel's index computation.

## Additional observations

- All compile modes affected: `default`, `reduce-overhead`, `max-autotune` (worst: diff up to 2.2e5)
- `float64` diffs range from 1e0 to 1e2 — not a precision issue
- Replacing `sum` with `mean` also triggers the bug
- `pool + sum(spatial_dims)` without `flatten` is correct — the flatten reshape changes the reduction structure
- Non-square spatial sizes (e.g., 8×10) also trigger
- Different pool output sizes (5, 7) all trigger under the same conditions

### Versions

- **PyTorch**: 2.13.0.dev20260420+cu126 (nightly)
- **CUDA**: 12.6
- **GPU**: Tesla T4
- **OS**: Ubuntu 20.04, Python 3.11

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #180898

Fixes #180848
Fixes #180956

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -2806,12 +2806,32 @@ def is_contiguous_storage_and_layout(x: IRNode) -> bool:
...
-            layout.pad_strides()
+            assert isinstance(layout, FlexibleLayout), type(layout)
+            layout = FixedLayout(
+                layout.device,
+                layout.dtype,
+                layout.size,
+                layout._pad_strides(layout.stride, layout.size, layout.dtype),
+                layout.offset,
+                layout.is_pinned,
+            )
```
Comment (shunting314) [inline torch/_inductor/ir.py:2817]: is the issue that the layout argument's stride get changed by this function?

Comment (jansel) [reply torch/_inductor/ir.py:2817]: Yeah, and that was hiding some other issues.

**Review by shunting314 — APPROVED** [review]
